pkg/urbit/noun/events.c

```diff
@@ -638,18 +635,123 @@ _ce_image_sync(u3e_image* img_u)
+//! |                  |   |  mapping
+//! |------------------|   |
+//! |                  |   |
+//! |                  | <-+
+//! +==================+ <- previous end of image in memory
+//! |                  |
+//! |                  |
+//! |      .....       |
+//! |                  |
+//! |                  |
+//! +==================+ <- base of image in memory
+//!  <low address>
+//! ```
+//!
+//! It's unclear whether anonymous mappings that lie within the previous bounds
+//! of the image (identified by `.....` in the diagram above) should be replaced
+//! with file-backed mappings when the image is resized. If the
+//! (Linux/macOS/Windows) kernel is smart enough to simply replace the
+//! anonymous mapping in the page cache with the file-backed mapping, then the
+//! anonymous mappings should be replaced with file-backed mappings. If, on the
+//! other hand, the kernel treats the pages of the two mappings as separate
+//! entities, then the cost of a cache miss may make replacing the anonymous
+//! mapping too expensive. For now, we remain conservative and don't replace the
+//! mappings and accept the potential cost of the accumulation of anonymous
+//! mappings (and the use of swap space that likely accompanies those mappings).
+//!
+//! @param[in] img_u  Image.
+//! @param[in] pgs_w  New size of the image.
+//! @param[in] bas_y  Base address of the image in memory. Used to establish a
+//!                   new mapping in memory. Should be NULL if no new mappings
+//!                   should be created.
 static void
-_ce_image_resize(u3e_image* img_u, c3_w pgs_w)
+_ce_image_resize(u3e_image* img_u, c3_w pgs_w, c3_y* bas_y)
 {
+  // The image is mapped into memory at base address `bas_y`.
+  if ( bas_y ) {
+    c3_y* ptr_y  = bas_y + c3_min(img_u->pgs_w, pgs_w) * pag_siz_i;
+    c3_ws dif_ws = (img_u->pgs_w - pgs_w) * pag_siz_i;
+
+    c3_i fla_i, fid_i, pro_i;
+    size_t off_i;
+    // The image shrunk.
+    if ( dif_ws > 0 ) {
+      fla_i = MAP_ANONYMOUS | MAP_FIXED | MAP_PRIVATE;
+      fid_i = -1;
+      pro_i = PROT_READ | PROT_WRITE;
+      off_i = 0;
+    }
+    // The image grew.
+    else if ( dif_ws < 0 ) {
+      fla_i = MAP_FIXED | MAP_PRIVATE;
+      fid_i = img_u->fid_i;
+      pro_i = PROT_READ;
+      off_i = (size_t)(ptr_y - bas_y);
+    }
+
+    if ( dif_ws != 0 && -1 == (c3_ps)mmap(ptr_y,
+                                          c3_abs(dif_ws),
+                                          pro_i,
+                                          fla_i,
+                                          fid_i,
+                                          off_i) )
+    {
+      fprintf(stderr,
+              "loom: failed to establish new mapping "
+              "for %s at %p after resizing: %s\r\n",
+              img_u->nam_c,
+              ptr_y,
+              strerror(errno));
+      exit(1);
+    }
+
+    if ( dif_ws < 0 ) {
+      off_i += pag_siz_i;
+    }
+    ptr_y += pag_siz_i;
```

Comment:
These pointer/offset calculations are unnecessary, left over from the per-page mapping loop.